airbyte_cdk/sources/declarative/incremental/concurrent_partition_cursor.py

```diff
@@ -325,7 +326,11 @@ def observe(self, record: Record) -> None:
                 "Invalid state as stream slices that are emitted should refer to an existing cursor"
             )
         self._cursor_per_partition[
-            self._to_partition_key(record.associated_slice.partition)
+            self._to_partition_key(
+                record.associated_slice.stream_slice.partition
+                if isinstance(record.associated_slice, AsyncPartition)
```

Comment:
It seems like a bad smell to have to check the instance of the stream slice. Do we actually need this? I thought that the records were created [during the record selection](https://github.com/airbytehq/airbyte-python-cdk/blob/3af96dcd713761d56190cb485f9376c65b1279c6/airbyte_cdk/sources/declarative/extractors/record_selector.py#L111) which would take the stream slice we pass [here](https://github.com/airbytehq/airbyte-python-cdk/blob/3af96dcd713761d56190cb485f9376c65b1279c6/airbyte_cdk/sources/declarative/retrievers/async_retriever.py#L95-L100). Is it just a matter of passing the right stream slice? 

The same comment applies to https://github.com/airbytehq/airbyte-python-cdk/pull/286/files#diff-6c148ab36dce37edc24313b91039c3122790e513b1cbba9bc3a4970460e37592R149